auctions/dutch-auction/src/main.sw

```diff
@@ -0,0 +1,280 @@
+
+        // Checks for correct asset_id being sent
+        require(msg_asset_id() == auction.asset_id, UserError::WrongAssetSent);
+
+        let price = calculate_price(auction_id);
+
+        // Checks for high enough amount being sent
+        require(price <= msg_amount(), UserError::BidTooLow);
+
+        // Disallows furthur bids
+        auction.ended = true;
+        storage.auctions.insert(auction_id, auction);
+
+        // If someone sends more than the current price, refunds the extra amount
+        if msg_amount() > price {
+            let return_amount = msg_amount() - price;
+            transfer_to_identity(return_amount, auction.asset_id, sender_indentity());
+        }
+
+        on_win(auction, price);
+
+        /// WARNING: This needs to be changed to a pop to a vec instead of just replacing the contents of the array
+        storage.active_auctions_by_identity.insert(auction.beneficiary, [0]);
+
+        log(WinningBidEvent {
+            winner: sender_indentity(), id: auction_id, 
+        });
+    }
+
+    /// Creates a new auction
+    /// Panics when -
+    ///     1. reserve_price is greater than opening_price
+    ///     2. current block height is higher than end_time or start_time
+    ///     3. start_time is greater than end_time
+    fn create_auction(opening_price: u64, reserve_price: u64, start_time: u64, end_time: u64, beneficiary: Identity, asset: ContractId) {
+        require(reserve_price <= opening_price, SetupError::EndPriceCannotBeLargerThanStartPrice);
+        require(height() < end_time, SetupError::AuctionCannotEndInThePast);
+        require(height() <= start_time, SetupError::AuctionCannotStartInThePast);
+        require(start_time < end_time, SetupError::AuctionCannotEndBeforeItStarts);
+
+        let auction = Auction {
+            opening_price, reserve_price, start_time, end_time, beneficiary, asset_id: asset,
+            ended: false,
+        };
+
+        storage.auction_count = storage.auction_count + 1;
+        storage.auctions.insert(storage.auction_count, auction);
+        /// WARNING: This needs to be changed to a push to a vec instead of just replacing the contents of the array
+        storage.active_auctions_by_identity.insert(beneficiary, [storage.auction_count]);
+
+        log(CreatedAuctionEvent {
+            id: storage.auction_count, auction, 
+        });
+    }
+
+    /// Cancels an auction so no one can bid on it.
+    /// Panics when -
+    ///     1. auction_id is 0 or higher than storage.auction_count
+    ///     2. msg_sender is not the beneficiary of the auction
+    ///     3. auction has already ended
+    fn cancel_auction(auction_id: u64) {
+        validate_id(auction_id);
+
+        let mut auction = storage.auctions.get(auction_id);
+
+        // Only the beneficiary can end the auction (prematurely)
+        require(eq_identity(sender_indentity(), auction.beneficiary), UserError::SenderNotBeneficiary);
```

Comment:
I think I may have said to switch this from the creator of the auction to the beneficiary. I've been unsure about that.
Having thought through a part of the user interface it makes sense to go back to the creator of the auction being the only one who can cancel it. I have some preliminary thoughts in the fundraiser spec